Src/FluentAssertions/Equivalency/CyclicReferenceDetector.cs

```diff
@@ -36,7 +36,7 @@ public bool IsCyclicReference(ObjectReference reference, Reason reason = null)
 
             if (reference.IsComplexType)
             {
-                if (orderedReferences.Contains(reference))
+                if (orderedReferences.TryGetValue(reference.Object, out var existingReference) && existingReference.Equals(reference))
```

Comment:
I think you have implement double comparison. This was not required before and should still not be required.
I think you should simply replace `List` by `HashSet`.

Reply:
That's fair. I was under the mistaken impression that a `Dictionary<object, T>` was doing a different type of comparison, but if it is simply using `.Equals()` and `.GetHashCode()` then this is indeed redundant.

Reply:
Oh hold on a second. Just looked at this more closely, and the subjects differ.

* When checking the dictionary, it's based on `reference.Object`, i.e. the target of the reference.
* When checking the equality, it's on `reference` itself. This comparison only returns true if the first reference is an earlier point in the chain of that led to the second reference.

This means that it can encounter the same object multiple times in an object graph without issue; each subsequent time it sees the object, it'll find it in the dictionary, but `existingReference.Equals(reference)` will say `false` -- even though it's the same object, the path it took to find `reference` didn't "come from" `existingReference`.

The code includes logic that replaces the dictionary entry when it sees the same object but it wasn't a circular reference. Just thinking about it logically, I'm thinking this may be unnecessary. The caller is using a depth-first search, and so if a circular reference existed, it would _always_ find that before it found unrelated references.

As an aside, when I just took a closer look at `ObjectReference.Equals`, I saw some semantics that seem dubious. The `a.Equals(b)` comparison includes a requirement that `a` be the "parent of" `b`. This would seem to me to break the invariant that `a.Equals(b) == b.Equals(a)`. I will rename `Equals` to something that conveys more clearly what it is doing.

Reply:
I'm not sure whether I fully understand your explanation.
I still think, `HashSet` should be sufficient. `List.Contains` and `HashSet.Contains` should have similar behavior, but `HashSet` is even faster because using a hash.

You changed the behavior. This would imply you have identified a new bug. Isnt't it?

Reply:
Nope, I did not change the behaviour. Comparing if the objects are the same is not the entire test. The dictionary entry is not the object itself, but the `ObjectReference`, and the test on the `ObjectReference`, which was (in my opinion incorrectly) named `Equals` but which I have now renamed, incorporates testing the path at which the object was encountered. For instance, if in the course of walking the object graph, it finds the same object first at `.A.Foo` and then subsequently at `B.Foo`, these objects are the same and will be the same dictionary entries, but they do not constitute a circular reference. But, if it encounters the same object first at `.A.Foo` and then subsequently at `.A.Foo.Bar.Qux.Foo`, then that _is_ a circular reference. The dictionary serves to identify the `ObjectReference` from `.A.Foo` (the previous occurrence) in both cases. The subsequent call to `ObjectReference.IsParentReferenceToSameObjectAs` differentiates these two cases.

This semantic is not new -- my change is only to use a dictionary to find the previous reference. Before my changes, instead of a dictionary it was a `List<ObjectReference>` upon which it did a linear scan.

Reply:
(I have come to understand this code better than when I was trying to explain it previously. :-) )

Reply:
The distinction between scanning the list and doing `IsParentReferenceToSameObjectAs` in the original code is obfuscated by the fact that `IsParentReferenceToSameObjectAs` is implemented by overriding `Equals`. It is being implicitly called by `orderedReferences.Contains`.

The main line's `ObjectReference.Equals` is defined as:
```
return ReferenceEquals(@object, other.@object) && IsParentOf(other);
```

Reply:
The primary intention of this PR is improving performance.
So, to improve performance you should use `HashSet` instead of `List`.
Of you think we have to fix incorrect behavior and you NEED a `Dictionary`, then please provide unit test which fails with `List` and `HashSet` but completes with `Dictionary`.

Reply:
I'm kind of astonished by this position, given the intense focus on correctness in this project. The `Object.Equals` instance method is part of a pattern that has very clearly-documented rules, and the current implementation breaks them -- even if you can't find a place in the code currently to _observe_ this causing problems, it is incorrect at a fundamental level. The Framework says, `a.Equals(b) == b.Equals(a)`, and everything within the framework is built on that invariant, and currently, the code is breaking that rule. It's pure chance that `List.Equals` does its comparison in a direction that actually works. Microsoft would not be breaking any rules if they changed the implementation of `List.Equals` so that it asked the supplied object if it was equal to the object in the collection and not vice versa. The direction of this comparison is not contractual _because it doesn't have to be_, because of the rules `Object.Equals` is "guaranteed" to follow.

A `List.Equals` check is conceptually far simpler than a `HashSet.Contains` or `Dictionary.Contains`. This is why, when writing this, switching from `List` to a more complicated type, I felt a great deal of uncertainly relying on undocumented internal behaviour of that new type. Even with `List<T>`, the code is technically incorrect for calling `.Contains`, but could be correct if, instead of calling `.Contains`, FluentAssertions did the loop itself and explicitly called the comparison method. But, at a deeper level, it is fundamentally wrong for the object to be overriding `.Equals` with an implementation that is not symmetrical, and to then be relying on Framework library code written against the documented properties of `.Equals` that aren't being followed... Yes, you have unit tests that say it works in its current form, but I'm quite convinced that it is a serious flaw! I feel certain that if it were something that current code analysis tech were up to, StyleCop would make this a build error.

Do you _really_ want to keep it in this state? Do you want to simply test that the internal implementation of `HashSet` seems to be doing all of its `.Equals` calls in the same direction as `List` and then latch onto that undocumented behaviour? I mean, it's your project, and if that's really what you want, I can change it to do that, but it feels at odds with the fundamental principles of the entire project to me!

Reply:
One other way to remedy this, that off the top of my head I _think_ would still serve the purpose, would be to make `ObjectReference.Equals` symmetrical simply by allowing either reference to be a parent of the other. This would mean that the `.Equals` implementation followed the rules, and within CyclicReferenceDetector.cs, it should still behave exactly the same way, because it will never find a parent to a reference _after_ that reference.

Reply:
The class is an internal class. **I myself** think it is acceptable that `Equals` (and `GetHashCode`) are implemented as required for the internal use case. @dennisdoomen  and @jnyrup may decide whether the strict implementation is required or not.

Further I think that we can trust in the current implementation (of `List` and `HashSet`). Microsoft is very conservative with changing behavior.

Please note in your explanation that `HashSet` and `Dictionary` are currently do exactly what you have implemented manually:
1. Get hash code of `ObjectReference.object`. The implementation of `ObjectReference.GetHashCode` is returning hash code of object.
2. Check for all objects with (potentially) same hash code using `Equals`.

The implementation **may** fail if we will have two instances of `ObjectReference` with same object and different parents. I do not know if this can ever happen. It seems so far it has never happen.
If so, we have a bug we can validate with a new unit test.

Reply:
I'm pretty sure that can and does come up when you have multiple references to an object that are not cyclic.
```
public class Test
{
  public string A;
  public string B;
}
```
Consider an instance where `A` and `B` point at the same `string` object. In this case, when comparing `obj.A`, it gets added to the `CyclicReferenceDetector` with a path of `[A]`. Then, later, when comparing `obj.B`, it goes back to the `CyclicReferenceDetector` with an `ObjectReference` for the same object, but this time the path is `[B]`. The dictionary will contain the previous reference with path `[A]`, but `IsParentReferenceToSameObjectAs` will return `false`, because `[A]` is not a parent of `[B]`. I'm pretty sure this is doing exactly what it is supposed to do.

With the old `List` implementation, the list would end up with both `ObjectReference` instances in it. With the new implementation, it can only refer to one of them at a time, since it's using the object itself as the dictionary key, but it seems to me that it really doesn't matter which one it is comparing, because:

* If there _is_ a cyclic reference, it will have already found it during the depth-first recursion into the object.
* If there _isn't_ a cyclic reference, then it doesn't matter what path it gets to the object by, it's never going to find a cyclic reference.

As for your assertion that what's happening internally in `HashSet` or `Dictionary` is _exactly_ the same, well, it's not -- or rather, it's not _guaranteed_ to be, and the crux of it is that with the current implementation, it makes a big difference whether the internal implementation is using `dataInHashSet.Equals(callerProvidedValue)` or `callerProvidedValue.Equals(dataInHashSet)` -- and it _shouldn't_ make a difference, per the contractual definition of `object.Equals`. As I've written it, the dictionary look-up is doing a strict object reference equality check on the _target_ of the `ObjectReference` to find the dictionary entry (this leverages `RuntimeHelpers.GetHashCode` to ensure that it isn't using any object's own override of `GetHashCode`), and then is calling `objectReference.IsParentReferenceToSameObjectAs` on the result of that as a separate step.

Reply:
I have been exploring the possibility of returning it to an `Equals` based implementation by making `Equals` behave in a symmetrical manner. I've run into a snag, though; in order to add a unit test of this, `FluentAssertions.Specs` needs to be able to see the internals of `FluentAssertions`, but `FluentAssertions` is signed and `FluentAssertions.Specs` is not. So, sign `FluentAssertions.Specs`, right? Well, `FluentAssertions.Specs` uses a NuGet package called `Chill` (@dennisdoomen I see you're a contributor on this too?), and `Chill` isn't signed. The requirement that signed assemblies only expose their `internal`s to other signed assemblies has a bit of a ripple effect. :-/

Reply:
For what it's worth, using a `HashSet<ObjectReference>` with `ObjectReference` providing a symmetrical `Equals` is passing all tests. The only snag is that currently, to make it testable, `ObjectReference` becomes a `public` type.

Reply:
Another thought occurs to me. Is there any situation where cyclic reference detection is being used that _isn't_ a depth-first scan of the object graph? If not, then instead of an ever-increasing collection of `ObjectReference` objects, maybe this could be achieved by means of essentially a stack of objects, so that objects are only in the collection for the time during which they are relevant.

Each time the tree walking code starts work on an object, it gets an `IDisposable` object from `CyclicReferenceDetector`. The creation of this object adds the subject to a `HashSet<object>(new ReferenceEqualsComparer())`, and then when it gets `Dispose`d, the object is removed from the `HashSet`. Instead of having to worry about paths at all, then, the check for cyclic references consists entirely of checking if an object is already in the `HashSet` before setting up a new scope for it.

Pseudocode:
```
public class CyclicReferenceDetector
{
  class ReferenceScope : IDisposable
  {
    CyclicReferenceDetector _owner;
    object _subject;

    public ReferenceScope(CyclicReferenceDetector owner, object subject)
    {
      _owner = owner;
      _subject = subject;
      _owner.EnterScope(subject);
    }

    public void Dispose()
    {
      _owner?.ExitScope(_subject);
      _owner = null;
    }
  }

  public IDisposable BeginScope(object subject)
    => new ReferenceScope(this, subject);

  HashSet<object> _objectsInScope = new HashSet<object>(new ReferenceEqualsComparer());

  void EnterScope(object subject)
  {
    if (!_objectsInScope.Add(subject))
      throw new Exception("Cyclic reference detected");
  }

  void ExitScope(object subject)
  {
    _objectsInScope.Remove(subject);
  }
}

...

void ObjectsAreEquivalent(object result, object expectation)
{
  using (var scope = _cyclicReferenceDetector.BeginScope(result))
  {
    foreach (var member in GetMembersForComparison(result))
    {
      object resultChild = member.GetValue(result);
      object expectationChild = member.GetValue(expectation);

      if (IsComplexType(resultChild) && IsComplexType(expectationChild))
        ObjectsAreEquivalent(resultChild, expectationChild);
      else
        ValuesAreEquivalent(resultChild, expectationChild);
    }
  }
}
```

Obviously it would need to be a bit more involved in order to give a meaningful error that can show you the path at which the cycle occurred, but can you think of a reason why this approach wouldn't work? It would mean that the `CyclicReferenceDetector` did not build up an ever-growing collection of objects that never shrank.

Reply:
Gee guys. I can barely catch up with all these long and in-depth discussions ;-)

Looking at the current code, I don't think the order matters because the parent condition check happens on the dotted path. However, [5 years ago],(https://github.com/fluentassertions/fluentassertions/commit/030b14c299c49c8b6047c2d79946c7e655a8496c), I did fix some unclear bug by switching from a `HashSet` to a simple list. Unfortunately, I did not add a test for this. Not sure why. So it _seems_ that the order _is_ important after all. 

But if we can't find any proof for that, then an implementation that first tries to find the `ObjectReference` based on a reference equality on the subject under test, and _then_ uses the `Equals` of `ObjectReference` may work indeed. I also agree with the dubious implementation of that `Equals`, so maybe renaming it to a more self-describing method makes sense.

Reply:
> an implementation that first tries to find the `ObjectReference` based on a reference equality on the subject under test, and _then_ uses the `Equals` of `ObjectReference`

This is what the (current) implementation of `HashSet.Contains` together with `ObjectReference` does.

Reply:
Yes, I think that should be fine.

Reply:
@lg2de It isn't, though, because the actual call to `.Equals` is taking place inside opaque framework code.

I am investigating making `.Equals` symmetric, though, which would obviate the entire issue. That's actually stuck on the `Chill` dependency downstream not being strong-named, though. :-P

Reply:
Okay, I believe this should be resolved to everyone's satisfaction now :-)

* ObjectReference.cs now uses `Equals` again, but this is okay because `Equals` is now symmetric. This is unit tested.
* CyclicReferenceDetector.cs can then confidently use `HashSet<ObjectReference>`. All existing tests pass.

I had thus far been amending my commit to incorporate PR review, having observed this elsewhere in the repository, but in this case the changes required for this are in a separate commit on the branch so they can be easily inspected independently.

Reply:
It does change the behavior of the original `ObjectReference`, but IMHO it does not matter whether two objects are each other's parent or child.